core/src/impl/Kokkos_HostSharedPtr.hpp

```diff
@@ -0,0 +1,182 @@
+
+ public:
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
+      MaybeReferenceCountedPtr&& other) noexcept
+      : m_value_ptr(other.m_value_ptr), m_control(other.m_control) {
+    other.m_value_ptr = nullptr;
+    other.m_control   = nullptr;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
+      const MaybeReferenceCountedPtr& other) noexcept
+      : m_value_ptr(other.m_value_ptr), m_control(other.m_control) {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    if (m_control) Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      MaybeReferenceCountedPtr&& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr       = other.m_value_ptr;
+      other.m_value_ptr = nullptr;
+      m_control         = other.m_control;
+      other.m_control   = nullptr;
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      const MaybeReferenceCountedPtr& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr = other.m_value_ptr;
+      m_control   = other.m_control;
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+      if (is_reference_counted())
+        Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION ~MaybeReferenceCountedPtr() { cleanup(); }
+
+  KOKKOS_FUNCTION T* get() const noexcept { return m_value_ptr; }
+  KOKKOS_FUNCTION T& operator*() const noexcept {
+    KOKKOS_EXPECTS(bool(*this));
+    return *get();
+  }
+  KOKKOS_FUNCTION T* operator->() const noexcept { return get(); }
+
+  // checks if the stored pointer is not null
+  KOKKOS_FUNCTION explicit operator bool() const noexcept {
+    return get() != nullptr;
+  }
+
+  // checks whether the MaybeReferenceCountedPtr manages the lifetime of the
+  // object
+  KOKKOS_FUNCTION bool is_reference_counted() const noexcept {
+    return m_control != nullptr;
+  }
+
+ private:
+  KOKKOS_FUNCTION void cleanup() noexcept {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    // If m_counter is set, then this instance is responsible for managing the
+    // objects pointed to by m_counter and m_value_ptr.
+    if (is_reference_counted()) {
+      KOKKOS_EXPECTS(m_control->m_deleter);
+      int const count = Kokkos::atomic_fetch_sub(&(m_control->m_counter), 1);
+      if (count == 1) {
+        (m_control->m_deleter)(m_value_ptr);
+        m_value_ptr = nullptr;
+        delete m_control;
+        m_control = nullptr;
+      }
+    }
+#endif
+  }
+
+  struct Control {
+    std::function<void(T*)> m_deleter;
+    int m_counter;
+  };
+
+  T* m_value_ptr;
+
+ protected:
+  Control* m_control;
+};
+
+template <class T>
+class HostSharedPtr : public MaybeReferenceCountedPtr<T> {
+ public:
+  explicit HostSharedPtr(T* value_ptr = nullptr)
+      : MaybeReferenceCountedPtr<T>(value_ptr, [](T* const t) { delete t; }) {}
+
+  template <class Deleter>
+  explicit HostSharedPtr(
```

Comment:
If we keep the default parameters in the single parameter constructor in line 160, then the ones here don't accomplish anything.  (I believe it is currently innocuous because this is a templated constructor and the one on line 160 is not.)

And do we really want the ability to set the deleter w a null ptr value?  That kind of adds a requirement that the `deleter` can handle a null ptr value.  (Fun fact:  this is one of the differences between `shared_ptr` deleters and `unique_ptr` deleters, and I never remember which is which.)

Reply:
Hmm... IMHO, it's more useful for `HostSharedPtr` and `UnmanagedPtr` to be semi-regular than to disallow assigning `nullptr` (in particular since `std::shared_ptr` also allows this).

Reply:
I am not objecting to and I agree that we should allow assigning a null pointer value (I use those words to differentiate from `nullptr`, because the latter has a different type).

Rephrasing my question from above:  whose responsibility is it to handle cleanup of a pointer with the null pointer value?

I believe that `MaybeReferenceCountedPtr` should be responsible, and the best way to accomplish this is by not creating a Control block at construction time if `element_ptr` has the null pointer value.  Its constructor should look something like:

```
template <class Deleter>
   MaybeReferenceCountedPtr(T* element_ptr, Deleter deleter)
       : m_element_ptr(element_ptr), m_control(nullptr) {
    if (element_ptr) {
     try {
       m_control = new Control{deleter, 1};
} catch (...) {
       deleter(element_ptr);
       throw;
     }
   }
}
```